presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -442,25 +447,48 @@ private synchronized QueryResults getNextResult(long token, UriInfo uriInfo, Str
         // last page is removed.  If another thread observes this state before the response is cached
         // the pages will be lost.
         Iterable<List<Object>> data = null;
+        List<String> binaryData = null;
         try {
-            ImmutableList.Builder<RowIterable> pages = ImmutableList.builder();
-            long bytes = 0;
             long rows = 0;
+            long bytes = 0;
             long targetResultBytes = targetResultSize.toBytes();
-            while (bytes < targetResultBytes) {
-                SerializedPage serializedPage = exchangeClient.pollPage();
-                if (serializedPage == null) {
-                    break;
+            if (binaryResults) {
+                binaryData = new ArrayList<>();
+                while (bytes < targetResultBytes) {
+                    SerializedPage serializedPage = exchangeClient.pollPage();
+                    if (serializedPage == null) {
+                        break;
+                    }
+
+                    rows += serializedPage.getPositionCount();
+                    bytes += serializedPage.getSizeInBytes();
+
+                    DynamicSliceOutput sliceOutput = new DynamicSliceOutput(1000);
+                    PagesSerdeUtil.writeSerializedPage(sliceOutput, serializedPage);
+
+                    String encodedPage = Base64.getEncoder().encodeToString(sliceOutput.slice().byteArray());
```

Comment:
Nit: Add a constant?

```suggestion
                    String encodedPage = BASE64_ENCODER.encodeToString(sliceOutput.slice().byteArray());
```